### ISSUE TITLE
Python tiledbsoma:  how to modify underlying schema for performance?

### ISSUE THREAD

**Issue by bkmartinjr** [single-cell-data/TileDB-SOMA#322]
Feature request:  the foundational data types (SOMADataFrame, SOMASparseNdArray, etc) should have a means by which advanced users can provide storage-engine specific schema, compatible with but different from the default schema. A default schema should always be available (and likely will be good enough for most applications), but there are examples where tuning is required.

The use cases I can thnk of, for a TileDB-backed SOMA implementation, are mostly about performance with different access patterns, or data that wants non-default different filters.  At a minimum, it would be beneficial if platform-specific hints (or just a tiledb schema) could be provided that would allow modification of:
* tile_order & cell_order
* tiling size: capacity, etc
* filter list for the array attributes

I don't think these can be done with native tiledb schema evolution post-creation, so this needs to be provided to the SOMA object-type-specific `create` methods in order to have it set upon the creation of the underlying array.

**Comment by johnkerl** [comment]
@bkmartinjr -- in the old API impl there's a goodly amount of this available:

https://github.com/single-cell-data/TileDB-SOMA/blob/main-old/apis/python/src/tiledbsoma/soma_options.py

* Not saying this is ideal, just a point of reference
* Maybe the configs there are enough -- we could port these as-is, en masse, to the new API impl
* Maybe we can add more such
* -- or -- note that as these are articulated now, they're single strings/ints which are parameters for passing to tiledbschema-create -- maybe in addition we could support putting more complex structures into a soma-options configuration -- ?

Open to your thoughts here on how to organize this kind of configuration control at the user <-> tiledbsoma interface/boundary

**Comment by bkmartinjr** [comment]
I have several thoughts, but I would also like @aaronwolen and @thetorpedodog to weigh in as I believe this is a broader API design pattern we will want to use in several other places.

Thoughts:
* the abstract spec already has the concept of `platform_config` extensions to any operation, but does not currently have any useful guidance beyond that...
* I would like to allow multiple platform-specific configs in a single bit of end-user code.  Ie, the ability of an SOMA user to embed platform-specific config for multiple storage engines (tiledb, HDF5, whatever...), without requiring the execution context have all of them installed (just need the one actually being used at any given moment)
* In this particular case (schema config), it seems like there are two different approaches (I prefer the first):
    1. add a bunch of declarative options as you have suggested above.  The only downside to this is that it requires that we come up with a syntax to encode more complicated config like filter chains.
    2. or, have the user create a schema, which the system then uses (or introspects and copies config out of), if it passes some basic compatibility tests.

Given all of this, I suggest a simple extension of your approach:
* for all operations, use  a single platform-specific config param, eg, `platform_config` pointing at a dict/map
* which contains top-level name/key that is the same as the string returned by the soma.get_storage_engine operation.  In the case of `tiledbsoma`, it would be `tiledb`
* which contains simple declarative config as you have defined above
I guess we could add another layer of hierarchy allowing values to be separated by operation, but I'm not sure that is necessary.

So a concrete example for SOMASparseNdArray.create:
```
a = SOMASparseNdArray(uri)
a.create(schema, platform_config = {
    "tiledb": {
        "capacity": 10_000,
        "attribute_filters": [
            { "name": "bitshuffle_filter" },
            { "name": "zstd_filter", "level": 22 },
        ]
    },
    "other_platform": { ... }
})
```

Thoughts?  I have no dogma here - just want something clear, expressive and able to support multiple platforms.  I am wondering if anyone has seen useful prior art elsewhere?

**Comment by bkmartinjr** [comment]
CC: @thetorpedodog - would love you to think about this as well. For several of our "challenging" use cases, I will definitely need the ability to tune the underlying schema of the NdArray & DataFrame objects. If we can agree on an approach, I could start pulling it through the API.

**Comment by thetorpedodog** [comment]
Sorry for missing this earlier—I’ll start thinking about it. Preliminarily, your solution seems to make a lot of sense.

**Comment by thetorpedodog** [comment]
One question I do have about the first line of your example (and which may branch into a deeper question)

```
a = SOMASparseNdArray(uri)
```

Is the intent that a user will be creating the array from the abstract interface, i.e. if they have, for example `tiledbsoma` installed, will they be calling `tiledbsoma.SparseNDArray(some_uri)` or `somabase.SparseNDArray(some_uri)`?

If the former, we need a way to disambiguate what backend implementation we're using. If both `tiledbsoma` and (hypothetically) `csvsoma` know how to store data in the filesystem, then `file:///path/to/my/somas` will not be enough information to read the data—it may be in TileDB format or in CSV format.

On the plus side, there is a way for a package to [register itself as a plugin](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins) given an extension point (which `somabase` could provide).

**Comment by bkmartinjr** [comment]
> Is the intent that a user will be creating the array from the abstract interface

No, we always intended that the user would first import a _concrete_ implementation that provides the SOMA API, then create the array with that.  So my example should read:

```
import tiledbsoma
a = tiledbsoma.SOMASparseNdArray(uri)
a.create(schema, platform_config = {
    "tiledb": {
        "capacity": 10_000,
        "attribute_filters": [
            { "name": "bitshuffle_filter" },
            { "name": "zstd_filter", "level": 22 },
        ]
    },
    "other_platform": { ... }
})
```

We discussed, and decided against, a pluggable "generic" SOMA implementation, as it appeared to be a far more complex effort than simply having separate & independent implementations sharing a common API.

I pointed you at this because I think we want some generic conventions that create a "safe space" for platform extensions/hooks, and because I'd like the `tiledbsoma` implementation of those hooks to be useful/easy-to-use.

**Comment by bkmartinjr** [comment]
Notes from design review meeting:
* Paul will propose something
* Needs:  a) array schema configuration, b) global tiledb configuration, c) context configuration (eg, set AWS credentials on read), etc.

**Comment by thetorpedodog** [comment]
I started working on [guidelines for what a generic `platform_config` object should look like](https://docs.google.com/document/d/1v7oe8fX86gM72Fg03t9DexNxITHhgaTaG-AwnUlSBPU/edit), but in doing so I have started thinking that maybe some kind of “SOMA Context” type might be called for. This is because from the note above, part (a) is per-write, but parts (b) and (c) are persistent and unchanging across an entire session with one dataset.

The getter operations (e.g. `.shape`) don't have a way to provide this per-platform config on each operation which may hit the storage backend, which means something like this needs to be kept around on each SOMA object which is tied to a backend dataset.

General thoughts:

As a rule, a Context object is created once by user code *before starting loading anything* and is not modified. It contains the global-level information we discuss. Any third-party code (e.g. a library which does stuff with SOMA objects) should treat it as an opaque token to be passed but not modified.

Each SOMA type which is directly backed by a storage engine concept has a `context` field. This can be propagated through object construction (e.g., [to use tiledb terminology] getting an array that lives in a group) by passing the context into the child so that it has access to all these global settings.

Two alternatives for what the context is:

1. Since a context is strictly for use by one storage engine, and we don't expect to be joining data from different "sessions" (i.e., everything that is used in one SOMA is stored in the same place), it can be a SOMA-implementation–defined type. `tiledb-soma` might have `TileDBSOMAContext`, which has TileDB-specific fields that are used by the `tiledb-soma` internals.
2. The context is essentially a "read only" layer of `platform_config` options. When an operation is performed, the per-operation `platform_config` (which provides ephemeral information, like how to do the current read) is layered on top of the Context's configuration (which provides semi-global information, like login credentials or data storage region).

**Comment by bkmartinjr** [comment]
@thetorpedodog - I like the `context` idea.  We will need to specify how it propagates, and if there is a global default context the user can manipulate.  For example, the Python SOMACollection currently auto-magically creates other SOMA objects in __getattr__ -- do these new objects inherit a context from the collection, from a global default, or?

A context also resolves the issue of the `tiledb.Ctx` - currently passed as a parameter on various functions, which is not as clean as I would like. Putting the `tiledb.Ctx` into a SOMAContext is a nice solution.

How do you see resolving per-call platform-specific config?  The simplest example is array-creation config, to be specified only in the `create()` methods?  These seem to still require a `platform_config`-like capability (although perhaps in some cases, they would be backstopped by a default in the SOMAContext?)

**Comment by bkmartinjr** [comment]
@thetorpedodog - thought I would add a real-world example of code I currently have, which would be ideally improved by the `Context`.

My typical "setup" for using SOMA currently looks like:
```
import tiledb
import tiledbsoma as soma

census_uri = "s3://.../"
census_region = "us-east-1"
config = {
    "py.init_buffer_bytes": 1024**3,
    "py.deduplicate": "true",
    "vfs.s3.region": census_region,
}
census = soma.SOMACollection(census_uri, ctx=tiledb.Ctx(config))
```

In an _ideal_ world, I would not have to import the `tiledb` package, or explicitly create the `tiledb.Ctx` object.  Instead, it would be nice if we had something approximating:

```
import tiledbsoma as soma

census_uri = "s3://.../"
census_region = "us-east-1"
context = soma.Context(platform_config={
    "tiledb": {
        # stuff in here is TileDB-specific (eg, can be handed directly to tiledb.Config)
        "py.init_buffer_bytes": 1024**3,
        "py.deduplicate": "true",
        "vfs.s3.region": census_region,
    },
    "some_other_engine": {
        # ...
    }
})
census = soma.SOMACollection(census_uri, context=context)
```

In other words, I don't have to import every storage engine API that I want to include a "config" for...

**Comment by bkmartinjr** [comment]
@johnkerl @thetorpedodog - I believe this issue is fully resolved, no? Can we close it if that is true?  I am successfully using the `platform_config` hook in `create` to modify the schema.

**Comment by thetorpedodog** [comment]
This is solved as far as the `platform_config` for creation goes. There is still future work for context integration through the codebase, but I think that is a separate issue than this one.
